File: backend/src/schemas/insights.py
"""Pydantic schemas for insights API."""

from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

# ============================================================================
# Validation Schemas
# ============================================================================


class ValidateInsightRequest(BaseModel):
    """Request to validate insight parameters before generation."""

    chat_ids: list[str] = Field(..., description="List of chat IDs to include")
    start_date: datetime
    end_date: datetime


class ValidateInsightResponse(BaseModel):
    """Response for insight validation."""

    valid: bool
    message_count: int
    exceeds_limit: bool = False
    estimated_tokens: int | None = None
    max_messages: int | None = None
    suggested_filters: dict[str, Any] | None = None


# ============================================================================
# Generate Schemas
# ============================================================================

# Supported languages for insight generation
SUPPORTED_LANGUAGES = {
    "en": "English",
    "zh-CN": "Chinese (Simplified)",
    "zh-TW": "Chinese (Traditional)",
    "ja": "Japanese",
    "ko": "Korean",
    "es": "Spanish",
    "fr": "French",
    "de": "German",
    "pt": "Portuguese",
    "ru": "Russian",
}

# Type-safe language code validation
SupportedLanguage = Literal["en", "zh-CN", "zh-TW", "ja", "ko", "es", "fr", "de", "pt", "ru"]


class GenerateInsightRequest(BaseModel):
    """Request to generate insights."""

    chat_ids: list[str] = Field(..., description="List of chat IDs to include")
    start_date: datetime
    end_date: datetime
    language: SupportedLanguage = Field(
        default="en", description="Language code for insight output"
    )


class GenerateInsightFromIdsRequest(BaseModel):
    """Request for generating an insight from specific message IDs."""

    message_ids: list[UUID]
    language: SupportedLanguage = "en"

    @field_validator("message_ids")
    @classmethod
    def validate_message_ids(cls, v):
        if len(v) == 0:
            raise ValueError("At least one message ID required")
        if len(v) > 10000:
            raise ValueError("Maximum 10000 message IDs per request")
        return v


# ============================================================================
# Insight Response Schemas
# ============================================================================


class InsightSummary(BaseModel):
    """Summary of an insight for list view."""

    id: UUID
    chat_ids: list[int]
    chat_titles: list[str] | None
    start_date: datetime
    end_date: datetime
    message_count: int
    status: str
    created_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class InsightDetail(BaseModel):
    """Full insight detail."""

    id: UUID
    user_id: UUID
    chat_ids: list[int]
    chat_titles: list[str] | None
    start_date: datetime
    end_date: datetime
    message_count: int

    # Generated content
    summary: str | None

    # Status
    status: str
    error_message: str | None

    # Metadata
    model_used: str
    provider_used: str
    input_tokens: int | None
    output_tokens: int | None
    cost_usd: Decimal | None
    generation_time_ms: int | None

    # Timestamps
    created_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class InsightListResponse(BaseModel):
    """Response for listing insights."""

    insights: list[InsightSummary]
    total: int


# ============================================================================
# Error Schemas
# ============================================================================
# All error handling done via FastAPI HTTPException

File: frontend/src/types/insights.ts
/**
 * LLM Insights Types
 * Types for consent management, usage tracking, SSE streaming, and insights
 */

// Validation types
// NOTE: chat_ids are strings (matching backend schema), dates are REQUIRED
export interface ValidationRequest {
  chat_ids: string[]
  start_date: string // ISO string, REQUIRED
  end_date: string // ISO string, REQUIRED
}

export interface ValidationResponse {
  valid: boolean
  message_count: number
  exceeds_limit: boolean
  estimated_tokens: number
  max_messages: number | null
  suggested_filters?: {
    narrower_date_range?: { start: string; end: string }
    fewer_channels?: string[]
  }
}

// Supported languages for insight generation
export const SUPPORTED_LANGUAGES = {
  en: 'English',
  'zh-CN': 'Chinese (Simplified)',
  'zh-TW': 'Chinese (Traditional)',
  ja: 'Japanese',
  ko: 'Korean',
  es: 'Spanish',
  fr: 'French',
  de: 'German',
  pt: 'Portuguese',
  ru: 'Russian',
} as const

export type SupportedLanguage = keyof typeof SUPPORTED_LANGUAGES

// Generation types
export interface GenerationRequest {
  message_ids: string[]
  language?: SupportedLanguage // optional, defaults to 'en'
}

export type SSEEventType = 'status' | 'chunk' | 'metadata' | 'error'

export interface SSEStatusEvent {
  type: 'status'
  status: 'pending' | 'generating' | 'completed'
  insight_id?: string
}

export interface SSEChunkEvent {
  type: 'chunk'
  content: string
  insight_id: string
}

export interface SSEMetadataEvent {
  type: 'metadata'
  tokens_used: number
  input_tokens: number | null
  output_tokens: number | null
  cost: number | null
  provider: string
  model: string
  generation_time_ms: number | null
}

export interface SSEErrorEvent {
  type: 'error'
  error: string
  detail: string
  reset_at?: string
  suggested_filters?: ValidationResponse['suggested_filters']
}

export type SSEEvent = SSEStatusEvent | SSEChunkEvent | SSEMetadataEvent | SSEErrorEvent

// Utility for displaying costs
export function formatCost(costUsd: string | null): string {
  if (!costUsd) return 'Free (cached)'
  const value = parseFloat(costUsd)
  if (value === 0) return 'Free'
  if (value < 0.01) return '<$0.01'
  return `$${value.toFixed(4)}`
}

// Utility for displaying token counts
export function formatTokens(input: number | null, output: number | null): string {
  if (input === null && output === null) return 'N/A'
  const total = (input || 0) + (output || 0)
  return `${total.toLocaleString()} tokens`
}

// Insight types - matches backend schema
// NOTE: Response types use number[] for chat_ids (backend returns integers)
export interface InsightSummary {
  id: string
  created_at: string
  chat_ids: number[] // RESPONSE: integers from backend
  chat_titles: string[] | null
  start_date: string | null
  end_date: string | null
  message_count: number
  status: 'pending' | 'generating' | 'completed' | 'failed'
  completed_at: string | null
}

export interface InsightDetail extends InsightSummary {
  user_id: string
  summary: string | null // Full markdown content
  error_message: string | null
  model_used: string
  provider_used: string
  input_tokens: number | null
  output_tokens: number | null
  cost_usd: string | null // Decimal string
  generation_time_ms: number | null
}

export interface InsightListResponse {
  insights: InsightSummary[]
  total: number
}

// Error types
export type InsightErrorCode =
  | 'rate_limit_exceeded'
  | 'message_limit_exceeded'
  | 'generation_failed'

// Custom error class for insight API errors
export class InsightApiError extends Error {
  constructor(
    public status: number,
    public data: {
      detail?: string
      error?: string
      reset_at?: string
      suggested_filters?: ValidationResponse['suggested_filters']
    }
  ) {
    super(data.detail || data.error || 'Insight API error')
    this.name = 'InsightApiError'
  }
}
